wikitools/errors.py

```diff
@@ -0,0 +1,68 @@
+import collections
+
+from wikitools import console
+
+
+class LinkError:
+    def pretty(self):
+        return f'{console.blue("Note:")} {repr(self)}'
+
+
+class LinkNotFound(LinkError, collections.namedtuple('LinkNotFound', 'location')):
+    """
+    An error indicating missing link: a text or binary file does not exist, and there is no such redirect.
+    """
+
+    location: str
+
+    def __repr__(self):
+        return f'"{self.location}" was not found'
+
+
+class BrokenRedirect(
+    LinkError,
+    collections.namedtuple('BrokenRedirect', 'location redirect_lineno redirect_destination')
+):
+    """
+    An error indicating broken redirect: an article from the redirect.yaml file does not exist.
+    """
+
+    location: str
+    redirect_lineno: int
+    redirect_destination: str
+
+    def __repr__(self):
+        return 'Broken redirect (redirect.yaml:{}: {} --> {})'.format(
+            self.redirect_lineno, self.location.lower(), self.redirect_destination
+        )
+
+
+class MissingReference(LinkError, collections.namedtuple('MissingReference', 'location')):
+    """
+    An error indicating that a reference-style link is missing its counterpart:
+    [link][link_ref] exists, but [link_ref]: /wiki/Path/To/Article does not.
+    """
+
+    location: str
+
+    def __repr__(self):
+        return f'No corresponding reference found for "{self.location}"'
+
+
+class MissingIdentifier(
+    LinkError,
+    collections.namedtuple('MissingIdentifier', 'path identifier translation_available')
+):
+    """
+    An error indicating that in another article there is no heading or any other object
+    that would produce #such-reference.
+    """
+
+    path: str
+    identifier: str
+    translation_available: bool
+
+    def __repr__(self):
+        return 'There is no heading or other object with identifier "{}" in "{}"{}'.format(
+            self.identifier, self.path, '.' if self.translation_available else ' (no translation available).'
```

Comment:
this is unnecessary as it will produce two full stops when no translation is available. come to think of it, I better remove both full stops for consistency and style

Reply:
ah i forgot to remove the other full stop after the parentheses to make it work, but i guess without any is fine too.